File: server-go/internal/handler/template.go
package handler

import (
	"github.com/gin-gonic/gin"
	"github.com/surveyking/server/internal/dto"
	"github.com/surveyking/server/internal/pkg/response"
	"github.com/surveyking/server/internal/service"
)

// TemplateHandler handles survey template API endpoints.
type TemplateHandler struct {
	svc *service.TemplateService
}

func NewTemplateHandler(svc *service.TemplateService) *TemplateHandler {
	return &TemplateHandler{svc: svc}
}

func (h *TemplateHandler) List(c *gin.Context) {
	var query dto.TemplateQuery
	if err := c.ShouldBindQuery(&query); err != nil {
		response.Fail(c, response.CodeError, err.Error())
		return
	}
	result, err := h.svc.List(query)
	if err != nil {
		response.Fail(c, response.CodeError, err.Error())
		return
	}
	response.OK(c, result)
}

func (h *TemplateHandler) Get(c *gin.Context) {
	var query dto.TemplateQuery
	if err := c.ShouldBindQuery(&query); err != nil {
		response.Fail(c, response.CodeError, err.Error())
		return
	}
	view, err := h.svc.Get(query)
	if err != nil {
		response.Fail(c, response.CodeError, err.Error())
		return
	}
	response.OK(c, view)
}

func (h *TemplateHandler) Create(c *gin.Context) {
	var req dto.TemplateRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		response.Fail(c, response.CodeError, err.Error())
		return
	}
	id, err := h.svc.Create(req)
	if err != nil {
		response.Fail(c, response.CodeError, err.Error())
		return
	}
	response.OK(c, id)
}

func (h *TemplateHandler) Update(c *gin.Context) {
	var req dto.TemplateRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		response.Fail(c, response.CodeError, err.Error())
		return
	}
	if err := h.svc.Update(req); err != nil {
		response.Fail(c, response.CodeError, err.Error())
		return
	}
	response.OK(c, nil)
}

func (h *TemplateHandler) Delete(c *gin.Context) {
	var req dto.TemplateRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		response.Fail(c, response.CodeError, err.Error())
		return
	}
	if err := h.svc.Delete(req.ID); err != nil {
		response.Fail(c, response.CodeError, err.Error())
		return
	}
	response.OK(c, nil)
}

func (h *TemplateHandler) ListCategory(c *gin.Context) {
	var query dto.CategoryQuery
	if err := c.ShouldBindQuery(&query); err != nil {
		response.Fail(c, response.CodeError, err.Error())
		return
	}
	categories, err := h.svc.ListCategories(query.Mode)
	if err != nil {
		response.Fail(c, response.CodeError, err.Error())
		return
	}
	response.OK(c, categories)
}

func (h *TemplateHandler) ListTag(c *gin.Context) {
	var query dto.TagQuery
	if err := c.ShouldBindQuery(&query); err != nil {
		response.Fail(c, response.CodeError, err.Error())
		return
	}
	tags, err := h.svc.ListTags(query.Mode, query.Category)
	if err != nil {
		response.Fail(c, response.CodeError, err.Error())
		return
	}
	response.OK(c, tags)
}

File: website/src/api/template.ts
import request from './index'
import type { TemplateQuery, TemplateView, TemplateRequest, CategoryView, TagView, CategoryQuery, TagQuery } from '@/types/template'
import type { PageResponse } from '@/types/api'

export const templateApi = {
  // List templates with pagination
  list(params: TemplateQuery): Promise<PageResponse<TemplateView>> {
    return request.get('/template/list', { params })
  },

  // Get single template by ID
  get(id: string): Promise<TemplateView> {
    return request.get('/template', { params: { id } })
  },

  // Create new template
  create(data: TemplateRequest): Promise<string> {
    return request.post('/template/create', data)
  },

  // Update existing template
  update(data: TemplateRequest): Promise<void> {
    return request.post('/template/update', data)
  },

  // Delete template
  delete(id: string): Promise<void> {
    return request.post('/template/delete', { id })
  },

  // List template categories
  listCategories(params: CategoryQuery): Promise<CategoryView[]> {
    return request.get('/template/category/list', { params })
  },

  // List template tags
  listTags(params: TagQuery): Promise<TagView[]> {
    return request.get('/template/tag/list', { params })
  }
}
